### PR TITLE
Add amqp reporter support

### PR DESCRIPTION
**Pull request by ValeryPiashchynski**

### CODE REVIEW

File: reporter/amqp/amqp.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	defaultRmqExchange   = "zipkin"
+	defaultExchangeKind  = "direct"
+)
+
+// rmqReporter implements Reporter by publishing spans to a RabbitMQ exchange
+type rmqReporter struct {
+	e        chan error
+	channel  *amqp.Channel
+	conn     *amqp.Connection
+	Exchange string
```
Comment (jcchavezs) [inline reporter/amqp/amqp.go]: I am curious about `Exchange` and `Queue` being exported.

File: reporter/amqp/amqp.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+// see https://github.com/openzipkin/zipkin/tree/master/zipkin-collector/rabbitmq
+// if want to change default routing key or exchange
+func Exchange(e string) ReporterOption {
+	return func(c *rmqReporter) {
+		c.Exchange = e
+	}
+}
+
+// Queue sets the Queue used to send messages
+func Queue(t string) ReporterOption {
```
Comment (jcchavezs) [inline reporter/amqp/amqp.go]: Is this `t` name for the variable right?

Comment (ValeryPiashchynski) [reply reporter/amqp/amqp.go]: > Is this `t` name for the variable right?

Yep, will rename to queue

File: reporter/amqp/amqp.go (hunk 1 above)
Comment (ValeryPiashchynski) [reply reporter/amqp/amqp.go]: > I am curious about `Exchange` and `Queue` being exported.

Done, I made it unexported

**Comment by basvanbeek** [comment]
@ValeryPiashchynski  thank you for your contribution, greatly appreciated.

If we are considering adding an amqp reporter to this core library, we should provide a solution that allows different encodings. 

Currently Zipkin officially supports 2 types of encodings for its native V2 collectors over various transports:
- json
- proto3

Since this is the core library for Zipkin in Go I require both encodings to be possible. I also like to hear from @openzipkin/core if there are any existing amqp reporter implementations in other languages they know of. If not, we can use Go to bootstrap one which allows for cross language interop in the future.

**Comment by ValeryPiashchynski** [comment]
> @ValeryPiashchynski thank you for your contribution, greatly appreciated.
> 
> If we are considering adding an amqp reporter to this core library, we should provide a solution that allows different encodings.
> 
> Currently Zipkin officially supports 2 types of encodings for its native V2 collectors over various transports:
> 
> * json
> * proto3
> 
> Since this is the core library for Zipkin in Go I require both encodings to be possible. I also like to hear from @openzipkin/core if there are any existing amqp reporter implementations in other languages they know of. If not, we can use Go to bootstrap one which allows for cross language interop in the future.

Thanks @basvanbeek for you comment, I will update this PR to support both protocols

**Comment by jcchavezs** [comment]
@ValeryPiashchynski did you have any time to continue this great work?

**Comment by ValeryPiashchynski** [comment]
@jcchavezs Yep, but I have some problems. By some reason, Zipkin doesn't read my protobuf spans from rabbitmq queue. All seems good with JSON, but for the protobuf, I still investigating the problem

**Comment by basvanbeek** [comment]
@ValeryPiashchynski, can you commit the code changes to allow for pluggable transports in this PR? If the approach seems good to us we could potentially merge the AMQP reporter with JSON serialization and deal with the proto once it works. It also allows us to validate your refactoring earlier on. 

Thanks so far for the excellent work!

**Comment by ValeryPiashchynski** [comment]
@basvanbeek Yep, sure. Will do today

**Comment by ValeryPiashchynski** [comment]
@basvanbeek It's ok at the moment. I have fixed all comments and using this feature in production now.
I will return to protobuf implementation in a separate branch

**Comment by basvanbeek** [comment]
@ValeryPiashchynski I guess the only API change would be the addition of a ReporterOption to switch serialization type? If so then I'll merge now.

**Comment by ValeryPiashchynski** [comment]
@basvanbeek Is there any objections to merging this PR?

**Comment by basvanbeek** [comment]
all tests failed so those need to be fixed... also I'm very busy at the moment so don't have time to look into it this week.

**Comment by jcchavezs** [comment]
I can have a look but for such a change I would like @basvanbeek to agree.

Den tir. 15. jan. 2019, 21:24 skrev Bas van Beek <notifications@github.com:

**Comment by jcchavezs** [comment]
I mean to approve it too.

Den tir. 15. jan. 2019, 21:28 skrev José Carlos Chávez <jcchavezs@gmail.com:

**Comment by jcchavezs** [comment]
Hi @ValeryPiashchynski are you still interested on this feature to be merged? could you please make the tests pass?

**Comment by ValeryPiashchynski** [comment]
Hey @jcchavezs , no, I am not interested already in this feature :( The tests fail I guess because we need to have a real instance of rabbitmq to test properly this integration, but in travis we don't have.

**Comment by adriancole** [comment]
maybe add it? https://docs.travis-ci.com/user/database-setup/#rabbitmq

it might be nice to make running against real instance optional for
local dev. regardless it should be running in CI

**Comment by jcchavezs** [comment]
@ValeryPiashchynski can I ask you how you overcome this issue? (Even if the
answer is that you moved to another tracer).

Thanks!

**Comment by ValeryPiashchynski** [comment]
> @ValeryPiashchynski can I ask you how you overcome this issue? (Even if the answer is that you moved to another tracer). Thanks!

@jcchavezs Sure, but, what issue are you talking about?

**Comment by jcchavezs** [comment]
I mean, how did you solve the lack of amqp? did you use another reporter or
moved to another project.

tor. 18. apr. 2019, 12:02 skrev 0xdev <notifications@github.com>:

**Comment by ValeryPiashchynski** [comment]
@jcchavezs Got u. So, I use a fork of zipkin-go with the code presented in this PR. I feel sorry abandoning this PR, but, I just have no enough time to fix ci/cd errors :(

**Comment by jcchavezs** [comment]
I see. Thanks for the feedback.

If there is a real user need (like your case) I can spend some time helping
out with this PR. I will check it next week.

tor. 18. apr. 2019, 12:16 skrev 0xdev <notifications@github.com>:

**Comment by ValeryPiashchynski** [comment]
@jcchavezs Appreciate that. That would be really good if you could help with that PR

**Comment by basvanbeek** [comment]
@jcchavezs @ValeryPiashchynski friendly ping... do the both of you have time to bring this to completion?

**Comment by ValeryPiashchynski** [comment]
@basvanbeek I can resume work on this PR. I am going to fix Travis issues till the end of the week.

**Comment by ValeryPiashchynski** [comment]
See https://github.com/openzipkin/zipkin-go/pull/130

**Pull request closed without merging**